src/FileIO.jl

```diff
@@ -0,0 +1,240 @@
+#           save_measurement(filename, measurement)
+#
+# loading follows the same structure
+# > Each level (beyond the outermost save) also has an argument `entryname::String`.
+# > Each save_something() should write a "VERSION" and a "type" to the file. The
+#   former simplifies updates down the line, the latter allows for dispatch.
+# > Each load_something should be dispatched on a type, e.g.
+#   `load_model(data, ::Type(HubbardModelAttractive))`
+
+
+"""
+    save(filename, mc; force_overwrite=false, allow_rename=true)
+
+Saves the given MonteCarlo simulation `mc` to a JLD-file `filename`.
+
+If `allow_rename = true` the filename will be adjusted if it already exists. If
+`force_overwrite = true` it will be overwritten. In this case a temporary backup
+will be created. If neither are true an error will be thrown.
+"""
+function save(filename, mc::MonteCarloFlavor; force_overwrite=false, allow_rename=true)
+    @assert endswith(filename, ".jld")
+
+    # handle ranming and overwriting
+    isfile(filename) && !force_overwrite && !allow_rename && throw(ErrorException(
+        "Cannot save because \"$filename\" already exists. Consider setting " *
+        "`allow_reanme = true` to adjust the filename or `force_overwrite = true`" *
+        " to overwrite the file."
+    ))
+    if isfile(filename) && !force_overwrite && allow_rename
+        filename = _generate_unqiue_JLD_filename(filename)
+    end
+
+    if force_overwrite
+        parts = splitpath(filename)
+        parts[end] = "." * parts[end]
+        temp_filename = _generate_unqiue_JLD_filename(joinpath(parts...))
+        mv(filename, temp_filename)
+    end
+
+    mode = isfile(filename) ? "r+" : "w"
+    file = jldopen(filename, mode)
+    write(file, "VERSION", 1)
+    save_mc(file, mc, "MC")
+    close(file)
+
+    if force_overwrite
+        rm(temp_filename)
+    end
+
+    return filename
+end
+
+# Something like
+# existing_file.jld -> existing_file_aJ3c.jld
+function _generate_unqiue_JLD_filename(filename)
+    isfile(filename) || return filename
+    # those map to 0-9,    A-Z,        a-z
+    x = rand([(48:57)..., (65:90)..., (97:122)...])
+    s = "_$(Char(x))"
+    filename = filename[1:end-4] * s * ".jld"
+    while isfile(filename)
+        x = rand([(48:57)..., (65:90)..., (97:122)...])
+        s = string(Char(x))
+        filename = filename[1:end-4] * s * ".jld"
+    end
+    filename
+end
+
+"""
+    load(filename)
+
+Loads a MonteCarlo simulation from the given JLD-file `filename`.
+"""
+function load(filename)
+    data = JLD.load(filename)
+    @assert data["VERSION"] == 1
+    load_mc(data)
+end
+
+
+"""
+    resume!(filename[; kwargs...])
+
+Resumes a Monte Carlo simulation from a savefile generated by `run!(mc)`. Takes
+the same keyword arguments as `run!`. Returns the simulation and the state
+returned by `run!`.
+
+See also: [`run!`](@ref)
+"""
+function resume!(filename; kwargs...)
+    data = JLD.load(filename)
+
+    @assert data["VERSION"] == 1
```

Comment:
These probably shouldn't be `@assert`s but rather, for example, `condition || throw(ErrorException("msg"))` with an appropriate exception type.

Reply:
As a rule of thumb, `@assert`s should never trigger (they might be disabled by a command line flag and turned into no-ops).